### DISCUSSION TITLE
Over-The-Air updates with Firebase?

### DISCUSSION THREAD

**Discussion by LunatiqueCoder** [invertase/react-native-firebase#6692] [Q&A]
Hello!

Thank you for this great module! 

We will use Firebase for different features and we were wondering if we could also use Over-The-Air updates with Firebase? I came across a medium post but I could only see instructions for Android and is not using `react-native-firebase`: https://medium.com/@soumyasethy/react-native-over-the-air-ota-updates-using-firebase-34fe13614101

What are your thoughts on this? Are there any examples or other materials?

**Comment by mikehardy** [comment]
I am only aware of Microsoft's AppCenter [CodePush](https://github.com/microsoft/react-native-code-push) product and Expo's OTA [updates](https://docs.expo.dev/versions/latest/config/app/#updates) as systems that most people use.

From my research here in the past the feature of OTA updates of the javascript bundle can be split into

1) bundle hosting / publishing
2) remote version catalog hosting / compatibility resolution / new version detection
3) remote bundle download / storage
4) user update prompting (forced updates, delayed updates, revoked updates maybe, etc)
5) local / on-device multi-bundle resolution / non-default bundle loading

Using firebase storage + remote-config for items 1-3 is interesting and does seem like a good fit for firebase + react-native
item 4 is still an issue, but it is a solved problem in AppCenter and Expo OTA I believe?

item 5 is not technically difficult and using local prefs (on android and iOS to store "name of current desired bundle") seems easy enough

That blog post looks interesting but I would categorize it as a "toy". There are a lot of edge cases that can cause update failures and handling them with fallbacks etc to make sure things are reliable is important, and it's best to use the other modules' work in that area if possible.

I wonder if it might make sense to code up a [Firebase Extension](https://firebase.google.com/products/extensions) that implements a custom expo-updates server? https://docs.expo.dev/distribution/custom-updates-server/

Looks like it brings in lots of Expo deps https://github.com/expo/expo/blob/sdk-47/packages/expo-updates/package.json so unsure how that would work in a project that does not use Expo otherwise :thinking:  - codepush does not seem to support third part hosting of manifests / bundles but perhaps that is a PR they would accept?

I would hate to reinvent all of the error handling + signing + version resolution + API work etc in the other modules. Any reasonable OTA bundle updater will eventually grow to handle all of these via feature requests and PRs and the duplicate work seems like a terrible malinvestment

microsoft version has cleaner deps to start with at least https://github.com/microsoft/react-native-code-push/blob/fcb347f01cdd0d5d83df104c6af2afdf07531431/package.json#L37-L45 + https://github.com/microsoft/code-push/blob/6d2c0e91b14a207d1a506289d6d139a21421fcee/package.json#L25-L32

So...The "easiest" thing may be to implement protocol compatibility that enables complete reuse of those existing native modules. That is a Firebase Extension that implements a REST API on top of firebase functions + storage that generates a CodePush-compatible manifest file + bundle downloader, if that's possible, and a PR that allows react-native-code-push to be pointed at that URL, or expo-updates for the same.
